File: act col/03/Exercise03.java
// =================================================================
//
// File: Example03.java
// Authors:
//          A01704696 Samuel Octavio Gonzalez Azpeitia
//          A01704052 Martín Adrian Noboa Monar
// Description: This file implements the enumeration sort algorithm 
//				using Java's Threads.
//
// Copyright (c) 2023 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;

public class Exercise03 extends Thread {
	private static final int SIZE = 100_000;
	private int array[], start, end;

	public Exercise03(int start, int end, int array[]) {
		this.array = array;
		// place your code here
        this.start = start;
        this.end = end;
	}

    public void doTask(){
        int b[] = new int[SIZE];
        int c[] = new int[SIZE];
        for (int i = start; i < end; i++) {
            for (int j = 0; j < SIZE; j++) {
                if (this.array[i] > this.array[j] || this.array[i] == this.array[j] && j < i) {
                    b[i] += 1;
                }
            }
        }
        for(int i = start; i < end; i++){
            c[b[i]] = this.array[i];
        }

        for(int i = start; i < end; i++){
            this.array[i] = c[i];
        }
    }

	public void run() {
		// place your code here
        doTask();

	}

	public static void main(String args[]) {
		int array[] = new int[SIZE];
		int aux[] = new int[SIZE];
		long startTime, stopTime;
		double elapsedTime;
		int blockSize;
		Exercise03 threads[];

		Utils.randomArray(array);
		Utils.displayArray("before", array);

		// place your code here
        blockSize = SIZE / Utils.MAXTHREADS;
        threads = new Exercise03[Utils.MAXTHREADS];

		System.out.printf("Starting...\n");
		elapsedTime = 0;
		for (int i = 0; i < Utils.N; i++) {
			System.arraycopy(array, 0, aux, 0, array.length);

			startTime = System.currentTimeMillis();

			// place your code here.
            for (int j = 0; j < threads.length; j++) {
                threads[j] = new Exercise03(j * blockSize, (j + 1) * blockSize, aux);

            }
            for (int j = 0; j < threads.length; j++) {
                threads[j].start();
            }
            for (int j = 0; j < threads.length; j++) {
                try {
                    threads[j].join();
                                        

                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

			stopTime = System.currentTimeMillis();

			elapsedTime += (stopTime - startTime);
		}
		Utils.displayArray("after", aux);
		System.out.printf("avg time = %.5f\n", (elapsedTime / Utils.N));
	}
}

File: act col/06/Exercise03.cpp
// =================================================================
//
// File: exercise03.cpp
// Author(s):
//      Martin Adrian Noboa Monar - A01704052
//      Samuel Octavio González Azpeitia - A01704696
// Description: This file contains the code that implements the
//				enumeration sort algorithm using the TBB 
//				technology. To compile:
//				g++ exercise03.cpp -o app -I/usr/local/lib/tbb/include -L/usr/local/lib/tbb/lib/intel64/gcc4.4 -ltbb
//
// Algorithm reference:
//	https://www.osys.se/Archive/Papers/parallel-sort/node3.html
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
//			Sin Threads		Con Threads		Speed Up
//			636.750  ms		 179.651 ms		3.54
// =================================================================

#include <iostream>
#include <iomanip>
#include <chrono>
#include <cstring>
#include <cmath>
#include <tbb/blocked_range.h>
#include <tbb/parallel_for.h>
#include "utils.h"

using namespace std;
using namespace std::chrono;
using namespace tbb;

#define SIZE 10000

// implement your code
class Exercise03
{
private:
	int *array,*b,*c,size;

public:
	Exercise03(int *a, int s) : array(a),size(s) { 
        b = new int[size];
        c = new int[size];
    }


    void operator() (const blocked_range<int> &r) const {
		for (int i = r.begin(); i != r.end(); i++) {
			for (int j = 0; j < size; j++) {
				if (array[i] > array[j] || array[i] == array[j] && j < i) {
					b[i] += 1;
				}
			}
            for(int i = 0; i < size; i++){
			    c[b[i]] = array[i];
            }

            for(int i = 0; i < size; i++){
                array[i] = c[i];
            }
		}
        
    }
};

int main(int argc, char* argv[]) {
	int *array,*aux;
	// These variables are used to keep track of the execution time.
	high_resolution_clock::time_point start, end;
	double timeElapsed;

	array = new int[SIZE];
	aux = new int[SIZE];
	random_array(array, SIZE);
	display_array("before", array);
	cout << "Starting...\n";
	timeElapsed = 0;
	for (int j = 0; j < N; j++) {
        memcpy(aux, array, sizeof(int) * SIZE);
		start = high_resolution_clock::now();
		// call the implemented function
        Exercise03 obj(array,SIZE);
        parallel_for(blocked_range<int>(0, SIZE),  obj);

		end = high_resolution_clock::now();
		timeElapsed += 
			duration<double, std::milli>(end - start).count();
	}
	display_array("after", array);
	cout << "avg time = " << fixed << setprecision(3) 
		 << (timeElapsed / N) <<  " ms\n";

	delete [] array;
	return 0;
}